### DISCUSSION TITLE
How to upload a file in the Ben Eater Unlock sketch?

### DISCUSSION THREAD

**Discussion by sirquackiii** [TomNisbet/TommyPROM#33] [Q&A] [answered]
I need to upload a BIN file to the EEPROM with the Ben Eater Arduino program, how could you do that?

**Comment by sirquackiii** [comment, marked as answer]
**After thinking about this a bit more, here's a theory on how it could happen:**
So I first write a python script that would convert the BIN file into a list, which could look something like this:
```python
with open(rom.bin, "rb") as file:
  for x in file:
     list.append(x)
```
And then I could copy and paste that into the Arduino file, and go through the list and write to the EEPROM. 
(The list would not have 32k items, but the Arduino Code would first fill the entire EEPROM with EAs and then go through the machine code.)

Please feel free to point out any flaws with this plan.

**Reply by TomNisbet** [reply]
That would work.  Ben's code does exactly that - it includes the EEPROM data as a C array in the arduino program.

You could write a python program to create the C data or just google for "bin to C array" and you'll find a bunch of converters.

One issue is that the Arduino has very limited RAM, so the data can't be very large.  You could write it in chunks or look into PROGMEM to store and retrieve the data from the arduino's flash instead.
